pkg/util/slices.go

```diff
@@ -48,3 +51,57 @@ func CombineUnique[T constraints.Ordered](a, b []T) []T {
 	}
 	return a
 }
+
+// Filter returns a new slice that only contains elements from collection that
+// satisfy predicate.
+//
+//	// Filter in place
+//	numbers = Filter(numbers, isEven)
+//	// Filter into a new slice
+//	odds := Filter(numbers, isEven)
+func Filter[T any](collection []T, predicate func(T) bool) []T {
+	i := 0
+	out := make([]T, len(collection))
+	for j := range collection {
+		if predicate(collection[j]) {
+			out[i] = collection[j]
+			i++
+		}
+	}
+	return slices.Clip(out[:i])
+}
+
+// Map returns a new slice containing the results of fn for each element within
+// collection. Usage:
+//
+//	Map([]int{1, 2, 3}, func(i int) int {
+//		return i
+//	})
+func Map[T, K any](collection []T, fn func(T) K) []K {
+	out := make([]K, len(collection))
+	for i, el := range collection {
+		out[i] = fn(el)
+	}
+	return out
+}
+
+// MapBy returns a map populated with keys and values returned by fn.
+// Usage:
+//
+//	// Construct a set.
+//	MapBy(numbers, func(i int) (int, struct{}) {
+//		return i, struct{}{}
+//	})
+//
+//	// Construct a map of numbers to their square.
+//	MapBy(numbers, func(i int) (int, int) {
+//		return i, i * i
+//	})
+func MapBy[T any, K comparable, V any](collection []T, fn func(T) (K, V)) map[K]V {
```

Comment:
Random q, does this name come from some place? I haven't heard of a function called `MapBy` before (contrast with `Filter` and `Map` which are very standard). I suppose I would prefer `MapFrom` as a name, as in "make a map from this collection"

Reply:
@JeffSwenson had suggested `AsMap`. `MapBy` just happens to be the thing I say to myself whenever I build a map from a slice, "map this by this key" which I realize is not the best statement 😅 

Updated to `MapFrom`!